client/lib/use-memoized-value/index.ts

```diff
@@ -0,0 +1,18 @@
+import { useRef } from 'react';
+
+export type Predicate< T > = ( prev: T | undefined, next: T ) => boolean;
+
+const defaultCompare = < T >( prev: T | undefined, next: T ): boolean => prev === next;
+
+export function useMemoizedValue< T >(
+	value: T,
+	compare: Predicate< T > = defaultCompare
+): T | undefined {
```

Comment:
The function always returns strictly `T`, the `| undefined` case is impossible because `value` is strictly `T`.

The TypeScript compiler probably doesn't agree with this because in theory, `compare( undefined, value )` can return `true`, but it practice such a comparison would always be `false`.